### PR TITLE
make kubectl factory composeable

### PR DESCRIPTION
**Pull request by deads2k**
Alternate resolution of https://github.com/kubernetes/kubernetes/pull/38524.

Currently, the kubectl factory cannot be cleanly composed because without polymorphism, any calls which delegate to other factory methods cannot injected.  We cannot reasonably predict everything a composer would want to override, so enumeration of individual "we think this field is important" function is untenable.  On the other hand, having a method registry func and attaching methods to it resulted in chaos before 1.5 and the cleaner interface.

This pull takes the approach of building the factory in "rings" of subfactories.  RingN relies on RingN-1 and the overall factory is a set of nested factories.  No function in a "ring" is allowed to reference a peer function, but it may reference a parent ring's function.  This allows us to easily compose one chain for raw kube, but an extender can simply wrap a particular ring with his custom handling of particular functions and then continue the chain as normal.  This allows customization of each individual function.

It turns out that we have three rings.
 1. discovery, negotiation, and no-dep functions
 1. object typing and type mapping
 1. stuff that relies on type mapping (builder)

This pull does nothing split apart the dependencies.  No behavior changes.  There's more cleanup that could be done (particularly in naming), but I'd like to defer that to a later step.

@kubernetes/sig-cli @fabianofranz @AdoHe this is going to be a pain to rebase, so quick reviews are appreciated.
@ncdc @smarterclayton

### CODE REVIEW

File: pkg/kubectl/cmd/util/ring0_factory.go [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+	clientConfig := optionalClientConfig
+	if optionalClientConfig == nil {
+		flags := pflag.NewFlagSet("", pflag.ContinueOnError)
+		clientConfig = DefaultClientConfig(flags)
+	}
+
+	return NewRing0FactoryFromDiscovery(clientConfig, &discoveryFactory{clientConfig: clientConfig})
+}
+
+// NewRing0FactoryFrom allows an external caller to substitute a different discoveryFactory
```
Comment (ncdc) [inline pkg/kubectl/cmd/util/ring0_factory.go]: NewRing0FactoryFromDiscovery

File: pkg/kubectl/cmd/util/ring1_factory.go [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/kubectl"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/registry/extensions/thirdpartyresourcedata"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/runtime/schema"
+)
+
+// Ring0Factory holds the first level of factory methods.  No method may refer
```
Comment (ncdc) [inline pkg/kubectl/cmd/util/ring1_factory.go]: Ring1Factory

Comment (ncdc) [inline pkg/kubectl/cmd/util/ring1_factory.go]: second level

**Comment by AdoHe** [comment]
Classic OOP language like Java has two types of polymorphism: compile time(static binding) and runtime(dynamic binding). IIUC we are going to enable runtime polymorphism, like method overriding.

**Comment by AdoHe** [comment]
```
RingN relies on RingN-1 and the overall factory is a set of nested factories.
```
does this mean RingN-1 is the parent  of RingN?

**Comment by AdoHe** [comment]
>> No function in a "ring" is allowed to reference a peer function, but it may reference a parent ring's function.

any specific thoughts on why restrict peer function calls?

File: pkg/kubectl/cmd/util/BUILD [hunk 3]
```diff
@@ -18,6 +18,9 @@ go_library(
         "factory.go",
         "helpers.go",
         "printing.go",
+        "ring0_factory.go",
```
Comment (AdoHe) [inline pkg/kubectl/cmd/util/BUILD]: just as @ncdc comments, index starts from 1.

File: pkg/kubectl/cmd/util/ring0_factory.go [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/registry/extensions/thirdpartyresourcedata"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/runtime/schema"
+	utilflag "k8s.io/kubernetes/pkg/util/flag"
+	"k8s.io/kubernetes/pkg/util/homedir"
+)
+
+// Ring0Factory holds the first level of factory methods.  No method may refer
+// to any peer factory method
```
Comment (AdoHe) [inline pkg/kubectl/cmd/util/ring0_factory.go]: I would like to demonstrate what first level factory methods include here? discovery, negotiation, and no-dep functions?

Comment (fabianofranz) [inline pkg/kubectl/cmd/util/ring0_factory.go]: It feels like discovery and negotiation should have a dedicated ring, would make the custom implementations cleaner and dedicated to the specific purpose of understanding the context it runs (server stuff). Also no-dep functions are less likely to need alternate implementations. Could be a follow-up though.

File: pkg/kubectl/cmd/util/factory.go [hunk 5]
```diff
@@ -309,581 +85,35 @@ func makeInterfacesFor(versionList []schema.GroupVersion) func(version schema.Gr
...
-
-	clientConfig := optionalClientConfig
-	if optionalClientConfig == nil {
-		clientConfig = DefaultClientConfig(flags)
-	}
-
-	clients := NewClientCache(clientConfig)
+	// this looks weird, but there's a reason.  In order for composers to be able to provide alternative factory implementations
+	// they need to provide low level pieces of *certain* functions so that when the factory calls back into itself
+	// it uses the custom version of the function.  Rather than try to enumerate everything that someone would wnat to override
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/util/factory.go]: s/wnat/want

**Comment by deads2k** [comment]
> any specific thoughts on why restrict peer function calls?

If `ring1.func1` called `ring1.func2`, then if I override `ring1.func2`, but not `ring1.func1`, then the non-overridden `ring1.func2` will be called.  This prevents clean composition.

File: pkg/kubectl/cmd/util/ring0_factory.go [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+		flags := pflag.NewFlagSet("", pflag.ContinueOnError)
+		clientConfig = DefaultClientConfig(flags)
+	}
+
+	return NewRing0FactoryFromDiscovery(clientConfig, &discoveryFactory{clientConfig: clientConfig})
+}
+
+// NewRing0FactoryFrom allows an external caller to substitute a different discoveryFactory
+// Which allows for the client cache to be built in ring0, but still rely on a custom discovery client
+func NewRing0FactoryFromDiscovery(clientConfig clientcmd.ClientConfig, discoveryFactory DiscoveryClientFactory) Ring0Factory {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/ring0_factory.go]: > Would you ever call this directly, so does this need to be exposed?

OpenShift would call this

Comment (fabianofranz) [inline pkg/kubectl/cmd/util/ring0_factory.go]: Yeah exactly.

File: pkg/kubectl/cmd/util/ring1_factory.go [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+// to any peer factory method
+type Ring1Factory interface {
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/util/ring1_factory.go]: Optionally the interfaces could go in `factory.go` as the file having the contracts, and dedicate `ringN_factory.go` to the implementations.

**Comment by fabianofranz** [comment]
I'd suggest to add your explanation in the PR comment to a doc, something under docs/devel or docs/design for example.

**Comment by fabianofranz** [comment]
Overall I like it. It still presupposes some level of predictability about what composers will want to override, but the ring levels plus relying on delegation to upstream for things you don't want to override in a level will streamline that.

**Comment by deads2k** [comment]
comments addressed.

File: pkg/kubectl/cmd/util/BUILD [hunk 8]
```diff
@@ -16,6 +16,9 @@ go_library(
...
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/util/BUILD]: I don't like the ring names.  It's an arbitrary abstraction that has nothing to do with the actual use of the code.  But the approach is fine.  Call it "core" or "client" or "printing".

**Comment by deads2k** [comment]
@fabianofranz comments address, ptal.  I have to open the factory doc pull in a different repo

File: pkg/kubectl/cmd/util/factory_client_access.go [hunk 9]
```diff
@@ -0,0 +1,592 @@
...
+
+	clientConfig := optionalClientConfig
+	if optionalClientConfig == nil {
+		clientConfig = DefaultClientConfig(flags)
+	}
+
+	return NewClientAccessFactoryFromDiscovery(flags, clientConfig, &discoveryFactory{clientConfig: clientConfig})
+}
+
+// NewRing0FactoryFrom allows an external caller to substitute a different discoveryFactory
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/util/factory_client_access.go]: NewClientAccessFactoryFromDiscovery

**Comment by fabianofranz** [comment]
There are a few follow-ups specially about what composes each ring but the idea looks good and improves composition, which was our original target. LGTM and we'll handle the follow-ups later.

**Comment by soltysh** [comment]
LGTM for me as well.

**Comment by deads2k** [comment]
@k8s-bot kubemark e2e test this

**Pull request merged**